tests/metagen/typegraphs/sample/ts_upload/client.ts

```diff
@@ -0,0 +1,878 @@
+  [K in keyof T]: K extends string ? K : never;
+}[keyof T];
+
+type JsonLiteral = string | number | boolean | null;
+type JsonObject = { [key: string]: Json };
+type JsonArray = Json[];
+type Json = JsonLiteral | JsonObject | JsonArray;
+
+type ErrorPolyfill = new (msg: string, payload: unknown) => Error;
+
+/* QueryGraph section */
+
+class _QueryGraphBase {
+  constructor(private typeNameMapGql: Record<string, string>) {}
+
+  /**
+   * Get the {@link GraphQLTransport} for the typegraph.
+   */
+  graphql(addr: URL | string, options?: GraphQlTransportOptions) {
+    return new GraphQLTransport(
+      new URL(addr),
+      options ?? {},
+      this.typeNameMapGql,
+    );
+  }
+}
+
+// -------------------------------------------------- //
+
+
+const nodeMetas = {
+  scalar() {
+    return {};
+  },
+  
+  RootUploadFn(): NodeMeta {
+    return {
+      ...nodeMetas.scalar(),
+      argumentTypes: {
+        file: "RootUploadFnInputFileFile",
+        path: "RootUploadFnInputPathRootUploadFnInputPathStringOptional",
+      },
+    };
+  },
+  RootUploadManyFn(): NodeMeta {
+    return {
+      ...nodeMetas.scalar(),
+      argumentTypes: {
+        prefix: "RootUploadManyFnInputPrefixRootUploadFnInputPathStringOptional",
+        files: "RootUploadManyFnInputFilesRootUploadManyFnInputFilesFileList",
+      },
+    };
+  },
+};
+export type RootUploadFnInputFileFile = File;
+export type RootUploadFnInputPathString = string;
+export type RootUploadFnInputPathRootUploadFnInputPathStringOptional = RootUploadFnInputPathString | null | undefined;
+export type RootUploadFnInput = {
+  file: RootUploadFnInputFileFile;
+  path?: RootUploadFnInputPathRootUploadFnInputPathStringOptional;
+};
+export type RootUploadManyFnInputPrefixRootUploadFnInputPathStringOptional = RootUploadFnInputPathString | null | undefined;
+export type RootUploadManyFnInputFilesFile = File;
+export type RootUploadManyFnInputFilesRootUploadManyFnInputFilesFileList = Array<RootUploadManyFnInputFilesFile>;
+export type RootUploadManyFnInput = {
+  prefix?: RootUploadManyFnInputPrefixRootUploadFnInputPathStringOptional;
+  files: RootUploadManyFnInputFilesRootUploadManyFnInputFilesFileList;
+};
+export type RootUploadFnOutput = boolean;
+
+
+export class QueryGraph extends _QueryGraphBase {
+  constructor() {
+    super({
+      "RootUploadFnInputFileFile": "root_upload_fn_input_file_file!",
+      "RootUploadFnInputPathRootUploadFnInputPathStringOptional": "String",
+      "RootUploadManyFnInputPrefixRootUploadFnInputPathStringOptional": "String",
+      "RootUploadManyFnInputFilesRootUploadManyFnInputFilesFileList": "[root_uploadMany_fn_input_files_file]!",
+    });
+  }
+    
+  upload(args: RootUploadFnInput | PlaceholderArgs<RootUploadFnInput>) {
+    const inner = _selectionToNodeSet(
+      { "upload": args },
+      [["upload", nodeMetas.RootUploadFn]],
+      "$q",
+    )[0];
+    return new MutationNode(inner, [[".file"]]) as MutationNode<RootUploadFnOutput>;
```

Comment:
I'd recommend putting this metadata, the object paths, on the NodeMeta objects. You might have functions that take file arguments and if you collect them here, they'll be duplicated.